arch/src/aarch64/gicv2.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{boxed::Box, result};
+
+use kvm_ioctls::{DeviceFd, VmFd};
+
+use super::gic::{Error, GICDevice};
+
+type Result<T> = result::Result<T, Error>;
+
+// Unfortunately bindgen omits defines that are based on other defines.
+// See arch/arm64/include/uapi/asm/kvm.h file from the linux kernel.
+const KVM_VGIC_V2_DIST_SIZE: u64 = 0x1000;
+const KVM_VGIC_V2_CPU_SIZE: u64 = 0x2000;
+
+/// Get the address of the GICv2 distributor.
+fn get_dist_addr() -> u64 {
+    super::layout::MAPPED_IO_START - KVM_VGIC_V2_DIST_SIZE
+}
+
+/// Get the size of the GIC_v2 distributor.
+fn get_dist_size() -> u64 {
+    KVM_VGIC_V2_DIST_SIZE
+}
+
+/// Get the address of the GIC_v2 CPU.
+fn get_cpu_addr() -> u64 {
+    get_dist_addr() - KVM_VGIC_V2_CPU_SIZE
+}
+
+/// Get the size of the GIC_v2 CPU.
+fn get_cpu_size() -> u64 {
+    KVM_VGIC_V2_CPU_SIZE
+}
+
+/// Represent a GIC v2 device
+pub struct GICv2 {
+    /// The file descriptor for the KVM device
+    fd: DeviceFd,
+
+    /// GIC device properties, to be used for setting up the fdt entry
+    properties: [u64; 4],
+}
+
+impl GICv2 {
+    pub fn new(vm: &VmFd) -> Result<Box<GICDevice>> {
```

Comment:
The code here is mostly the same for GICv3.

From just looking at the two functions the only gic-version-specific things are the `attributes`/`properties`.
I think this whole gic-builder code can be shared (within a trait or at least as a static function) and only implement the `attributes` or `properties` part in a gic-version-specific code.

Reply:
Had thought about it as well. I 'll look into it. Thanks a lot for the review!